Data Wrangling Section for IBM Coursework

In [1]:

import pandas as pd
import matplotlib.pyplot as plt



<h3>Download Data Frame from Website, add headers, create data frame and save to Excel file</h3>


In [7]:
#### Load Data and store in dataframe Auto_df

import pandas as pd

# Define the data path and headers
datapath = 'https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01'

headers = [
    "symboling", "normalized-losses", "make", "fuel-type", "aspiration", "num-of-doors", "body-style",
    "drive-wheels", "engine-location", "wheel-base", "length", "width", "height", "curb-weight", "engine-type",
    "num-of-cylinders", "engine-size", "fuel-system", "bore", "stroke", "compression-ratio", "horsepower",
    "peak-rpm", "city-mpg", "highway-mpg", "price"
]


In [14]:
### Open the Local Data File

datapath = '/Users/rudolphsurovcik/Library/CloudStorage/GoogleDrive-grs3@lehigh.edu/My Drive/Lehigh Fall 2024/Python Developments/Data Files for Python/Auto85.xlsx'

Auto_df = pd.read_excel(datapath, names = headers)
Auto_df.head(10)

# Check the data types of the columns
Auto_df.dtypes

# describe the data
Auto_df.describe()


symboling  wheel-base      length       width      height  \
count  205.000000  205.000000  205.000000  205.000000  205.000000   
mean     0.834146   98.756585  174.049268   65.907805   53.724878   
std      1.245307    6.021776   12.337289    2.145204    2.443522   
min     -2.000000   86.600000  141.100000   60.300000   47.800000   
25%      0.000000   94.500000  166.300000   64.100000   52.000000   
50%      1.000000   97.000000  173.200000   65.500000   54.100000   
75%      2.000000  102.400000  183.100000   66.900000   55.500000   
max      3.000000  120.900000  208.100000   72.300000   59.800000   

       curb-weight  engine-size  compression-ratio    city-mpg  highway-mpg  
count   205.000000   205.000000         205.000000  205.000000   205.000000  
mean   2555.565854   126.907317          10.142537   25.219512    30.751220  
std     520.680204    41.642693           3.972040    6.542142     6.886443  
min    1488.000000    61.000000           7.000000   13.000000    16.000000  
25%    2145.000000    97.000000           8.600000   19.000000    25.000000  
50%    2414.000000   120.000000           9.000000   24.000000    30.000000  
75%    2935.000000   141.000000           9.400000   30.000000    34.000000  
max    4066.000000   326.000000          23.000000   49.000000    54.000000

<h2 id="identify_handle_missing_values">Identify and handle missing values</h2>

<h3 id="identify_missing_values">Identify missing values</h3>
<h4>Convert "?" to NaN</h4>
In the car dataset, missing data comes with the question mark "?".
We replace "?" with NaN (Not a Number), Python's default missing value marker for reasons of computational speed and convenience. Here we use the function: 
 <pre>.replace(A, B, inplace = True) </pre>
to replace A by B.

In [6]:
import pandas as pd

# Define the data path and headers
datapath = 'https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01'

headers = [
    "symboling", "normalized-losses", "make", "fuel-type", "aspiration", "num-of-doors", "body-style",
    "drive-wheels", "engine-location", "wheel-base", "length", "width", "height", "curb-weight", "engine-type",
    "num-of-cylinders", "engine-size", "fuel-system", "bore", "stroke", "compression-ratio", "horsepower",
    "peak-rpm", "city-mpg", "highway-mpg", "price"
]

# Load data and store it in a DataFrame
Auto_df = pd.read_csv(datapath, names=headers)

# Display the first ten rows of the DataFrame
print(Auto_df.head(10))

# Define the path to save the Excel file
save_path = '/Users/rudolphsurovcik/Library/CloudStorage/GoogleDrive-grs3@lehigh.edu/My Drive/Lehigh Fall 2024/Python Developments/Data Files for Python/Auto85.xlsx'

# Save the DataFrame to an Excel file
Auto_df.to_excel(save_path, index=False)

URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self-signed certificate in certificate chain (_ssl.c:1000)>

In [15]:
import numpy as np

# replace "?" to NaN
print(Auto_df.head(5))
Auto_df.replace("?", np.nan, inplace = True)
print(Auto_df.head(5))

   symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio hor

<h4>Evaluating for Missing Data</h4>

The missing values are converted by default. We use the following functions to identify these missing values. There are two methods to detect missing data:

<ol>
    <li><b>.isnull()</b></li>
    <li><b>.notnull()</b></li>
<ol>    
</ol>
The output is a boolean value indicating whether the value that is passed into the argument is in fact missing data.


In [16]:
### Check for missing data in Auto_df

missing_data = Auto_df.isnull() # True if missing data, False if not missing data
missing_data.head(5) # Display the first 5 rows of data frame with missing data


symboling  normalized-losses   make  fuel-type  aspiration  num-of-doors  \
0      False               True  False      False       False         False   
1      False               True  False      False       False         False   
2      False               True  False      False       False         False   
3      False              False  False      False       False         False   
4      False              False  False      False       False         False   

   body-style  drive-wheels  engine-location  wheel-base  ...  engine-size  \
0       False         False            False       False  ...        False   
1       False         False            False       False  ...        False   
2       False         False            False       False  ...        False   
3       False         False            False       False  ...        False   
4       False         False            False       False  ...        False   

   fuel-system   bore  stroke  compression-ratio  horsepower  peak-rpm  \
0        False  False   False              False       False     False   
1        False  False   False              False       False     False   
2        False  False   False              False       False     False   
3        False  False   False              False       False     False   
4        False  False   False              False       False     False   

   city-mpg  highway-mpg  price  
0     False        False  False  
1     False        False  False  
2     False        False  False  
3     False        False  False  
4     False        False  False  

[5 rows x 26 columns]

missing_data = df.isnull()
missing_data.head(5)

<h4>Count missing values in each column</h4>
<p>
Using a for loop in Python, we can quickly figure out the number of missing values in each column. As mentioned above, "True" represents a missing value and "False" means the value is present in the dataset.  In the body of the for loop the method ".value_counts()" counts the number of "True" values. 
</p>

In [17]:
### Count missing values in each column

for column in missing_data.columns.values.tolist(): # Loop through each column
    print(column)
    print (missing_data[column].value_counts()) # Count the number of missing data in each column
    print("")    

symboling
symboling
False    205
Name: count, dtype: int64

normalized-losses
normalized-losses
False    164
True      41
Name: count, dtype: int64

make
make
False    205
Name: count, dtype: int64

fuel-type
fuel-type
False    205
Name: count, dtype: int64

aspiration
aspiration
False    205
Name: count, dtype: int64

num-of-doors
num-of-doors
False    203
True       2
Name: count, dtype: int64

body-style
body-style
False    205
Name: count, dtype: int64

drive-wheels
drive-wheels
False    205
Name: count, dtype: int64

engine-location
engine-location
False    205
Name: count, dtype: int64

wheel-base
wheel-base
False    205
Name: count, dtype: int64

length
length
False    205
Name: count, dtype: int64

width
width
False    205
Name: count, dtype: int64

height
height
False    205
Name: count, dtype: int64

curb-weight
curb-weight
False    205
Name: count, dtype: int64

engine-type
engine-type
False    205
Name: count, dtype: int64

num-of-cylinders
num-of-cylinders
False    205
Nam

Based on the summary above, each column has 205 rows of data and seven of the columns containing missing data:

<ol>
    <li>"normalized-losses": 41 missing data</li>
    <li>"num-of-doors": 2 missing data</li>
    <li>"bore": 4 missing data</li>
    <li>"stroke" : 4 missing data</li>
    <li>"horsepower": 2 missing data</li>
    <li>"peak-rpm": 2 missing data</li>
    <li>"price": 4 missing data</li>
</ol>

<h3 id="deal_missing_values">Deal with missing data</h3>
<b>How to deal with missing data?</b>

<ol>
    <li>Drop data<br>
        a. Drop the whole row<br>
        b. Drop the whole column
    </li>
    <li>Replace data<br>
        a. Replace it by mean<br>
        b. Replace it by frequency<br>
        c. Replace it based on other functions
    </li>
</ol>

Whole columns should be dropped only if most entries in the column are empty. In our dataset, none of the columns are empty enough to drop entirely.
We have some freedom in choosing which method to replace data; however, some methods may seem more reasonable than others. We will apply each method to many different columns:

<b>Replace by mean:</b>

<ul>
    <li>"normalized-losses": 41 missing data, replace them with mean</li>
    <li>"stroke": 4 missing data, replace them with mean</li>
    <li>"bore": 4 missing data, replace them with mean</li>
    <li>"horsepower": 2 missing data, replace them with mean</li>
    <li>"peak-rpm": 2 missing data, replace them with mean</li>
</ul>

<b>Replace by frequency:</b>

<ul>
    <li>"num-of-doors": 2 missing data, replace them with "four". 
        <ul>
            <li>Reason: 84% sedans is four doors. Since four doors is most frequent, it is most likely to occur</li>
        </ul>
    </li>
</ul>

<b>Drop the whole row:</b>

<ul>
    <li>"price": 4 missing data, simply delete the whole row
        <ul>
            <li>Reason: price is what we want to predict. Any data entry without price data cannot be used for prediction; therefore any row now without price data is not useful to us</li>
        </ul>
    </li>
</ul>


<h4>Calculate the mean value for the "normalized-losses" column </h4>

In [18]:
avg_norm_loss = Auto_df["normalized-losses"].astype("float").mean(axis=0)
print("Average of normalized-losses:", avg_norm_loss)

Average of normalized-losses: 122.0


<h4>Replace "NaN" with mean value in "normalized-losses" column</h4>

In [29]:
import numpy as np

# Replace NaN values in the 'normalized-losses' column with avg_norm_loss

Auto_df["normalized-losses"] = Auto_df["normalized-losses"].replace(np.nan, avg_norm_loss)

<h4>Calculate the mean value for the "bore" column</h4>

In [23]:
#### Calculate the mean value for the 'bore' column

avg_bore=Auto_df['bore'].astype('float').mean(axis=0) # Calculate the mean value for the 'bore' column
print("Average of bore:", avg_bore) # Print the mean value for the 'bore' column

Average of bore: 3.3297512437810943


In [25]:
import pandas as pd

def clean_data(Auto_df):
    # One-hot encode column: 'make'
    insert_loc = Auto_df.columns.get_loc('make')
    Auto_df = pd.concat([Auto_df.iloc[:,:insert_loc], pd.get_dummies(Auto_df.loc[:, ['make']]), Auto_df.iloc[:,insert_loc+1:]], axis=1)
    return Auto_df

Auto_df_clean = clean_data(Auto_df.copy())
Auto_df_clean.head()





symboling normalized-losses  make_alfa-romero  make_audi  make_bmw  \
0          3             122.0              True      False     False   
1          3             122.0              True      False     False   
2          1             122.0              True      False     False   
3          2               164             False       True     False   
4          2               164             False       True     False   

   make_chevrolet  make_dodge  make_honda  make_isuzu  make_jaguar  ...  \
0           False       False       False       False        False  ...   
1           False       False       False       False        False  ...   
2           False       False       False       False        False  ...   
3           False       False       False       False        False  ...   
4           False       False       False       False        False  ...   

   engine-size  fuel-system  bore  stroke  compression-ratio  horsepower  \
0          130         mpfi  3.47    2.68                9.0         111   
1          130         mpfi  3.47    2.68                9.0         111   
2          152         mpfi  2.68    3.47                9.0         154   
3          109         mpfi  3.19    3.40               10.0         102   
4          136         mpfi  3.19    3.40                8.0         115   

   peak-rpm  city-mpg  highway-mpg  price  
0      5000        21           27  13495  
1      5000        21           27  16500  
2      5000        19           26  16500  
3      5500        24           30  13950  
4      5500        18           22  17450  

[5 rows x 47 columns]

In [27]:
"""
Cell generated by Data Wrangler.
"""
import pandas as pd

def clean_data(Auto_df):
    # One-hot encode column: 'make'
    insert_loc = Auto_df.columns.get_loc('make')
    Auto_df = pd.concat([Auto_df.iloc[:,:insert_loc], pd.get_dummies(Auto_df.loc[:, ['make']], dummy_na=True), Auto_df.iloc[:,insert_loc+1:]], axis=1)
    # Change column type to int8 for columns: 'make_alfa-romero', 'make_audi' and 21 other columns
    Auto_df = Auto_df.astype({'make_alfa-romero': 'int8', 'make_audi': 'int8', 'make_bmw': 'int8', 'make_chevrolet': 'int8', 'make_dodge': 'int8', 'make_honda': 'int8', 'make_isuzu': 'int8', 'make_mazda': 'int8', 'make_jaguar': 'int8', 'make_mercedes-benz': 'int8', 'make_mercury': 'int8', 'make_mitsubishi': 'int8', 'make_nissan': 'int8', 'make_peugot': 'int8', 'make_porsche': 'int8', 'make_plymouth': 'int8', 'make_renault': 'int8', 'make_saab': 'int8', 'make_subaru': 'int8', 'make_toyota': 'int8', 'make_nan': 'int8', 'make_volvo': 'int8', 'make_volkswagen': 'int8'})
    # Change column type to float64 for column: 'normalized-losses'
    Auto_df = Auto_df.astype({'normalized-losses': 'float64'})
    return Auto_df

Auto_df_clean = clean_data(Auto_df.copy())
Auto_df_clean.head()

symboling  normalized-losses  make_alfa-romero  make_audi  make_bmw  \
0          3              122.0                 1          0         0   
1          3              122.0                 1          0         0   
2          1              122.0                 1          0         0   
3          2              164.0                 0          1         0   
4          2              164.0                 0          1         0   

   make_chevrolet  make_dodge  make_honda  make_isuzu  make_jaguar  ...  \
0               0           0           0           0            0  ...   
1               0           0           0           0            0  ...   
2               0           0           0           0            0  ...   
3               0           0           0           0            0  ...   
4               0           0           0           0            0  ...   

   engine-size  fuel-system  bore  stroke  compression-ratio  horsepower  \
0          130         mpfi  3.47    2.68                9.0         111   
1          130         mpfi  3.47    2.68                9.0         111   
2          152         mpfi  2.68    3.47                9.0         154   
3          109         mpfi  3.19    3.40               10.0         102   
4          136         mpfi  3.19    3.40                8.0         115   

   peak-rpm  city-mpg  highway-mpg  price  
0      5000        21           27  13495  
1      5000        21           27  16500  
2      5000        19           26  16500  
3      5500        24           30  13950  
4      5500        18           22  17450  

[5 rows x 48 columns]

<h4>Replace "NaN" with the mean value in the "bore" column</h4>

In [30]:
#### Old version Auto_df["bore"].replace(np.nan, avg_bore, inplace=True)

Auto_df["bore"] = Auto_df["bore"].replace(np.nan, avg_bore)

from IPython.display import display, HTML

<div style="background-color: #cce5ff; color: #004085; padding: 20px; border-radius: 5px; margin-top: 20px; border: 1px solid #b8daff;">New Box Shown in Blue rather than Red.</div>'

In [33]:
#Calculate the mean value for "stroke" column

avg_stroke = round(Auto_df["stroke"].astype("float").mean(axis = 0),2)

print("Average of stroke:", avg_stroke) 



# replace NaN by mean value in "stroke" column

# Auto_df["stroke"].replace(np.nan, avg_stroke, inplace = True) <<<<Old version

Auto_df["stroke"] = Auto_df["stroke"].replace(np.nan, "stroke")

Average of stroke: 3.26
